src/diffusers/pipelines/stable_diffusion/safety.py

```diff
@@ -0,0 +1,94 @@
+        self.concept_embeds = self.get_text_embeds(
+            list(self.concepts_dict.keys()))
+        self.special_care_embeds = self.get_text_embeds(
+            list(self.special_care_dict.keys()))
+
+    def get_image_embeds(self, input):
+        """Get embeddings for images or tensor"""
+        with torch.cuda.amp.autocast():
+            with torch.no_grad():
+                # Preprocess if input is a list of PIL images
+                if isinstance(input, list):
+                    l = []
+                    for image in input:
+                        l.append(self.preprocess(image))
+                    img_tensor = torch.stack(l)
+                # input is a tensor
+                elif isinstance(input, torch.Tensor):
+                    img_tensor = input
+                return self.clip_model.encode_image(img_tensor.half().to(self.device))
+
+    def get_text_embeds(self, input):
+        """Get text embeddings for a list of text"""
+        with torch.cuda.amp.autocast():
+            with torch.no_grad():
+                input = open_clip.tokenize(input).to(self.device)
+                return(self.clip_model.encode_text(input))
+
+    def forward(self, images):
+        """Get embeddings for images and output nsfw and concept scores"""
+        image_embeds = self.get_image_embeds(images)
+        concept_list = list(self.concepts_dict.keys())
+        special_list = list(self.special_care_dict.keys())
+        special_cos_dist = pw_cosine_distance(image_embeds,
+                                              self.special_care_embeds).cpu().numpy()
+        cos_dist = pw_cosine_distance(image_embeds,
+                                      self.concept_embeds).cpu().numpy()
+        result = []
+        for i in range(image_embeds.shape[0]):
+            result_img = {
+                "special_scores":{},
+                "special_care":[],
+                "concept_scores":{},
+                "bad_concepts":[]}
+            adjustment = 0.05
+            for j in range(len(special_cos_dist[0])):
+                concept_name = special_list[j]
+                concept_cos = special_cos_dist[i][j]
+                concept_threshold = self.special_care_dict[concept_name]
+                result_img["special_scores"][concept_name] = round(
+                    concept_cos - concept_threshold + adjustment,3)
+                if result_img["special_scores"][concept_name] > 0:
+                    result_img["special_care"].append({concept_name,result_img["special_scores"][concept_name]})
+                    adjustment = 0.01
+            for j in range(len(cos_dist[0])):
+                concept_name = concept_list[j]
+                concept_cos = cos_dist[i][j]
+                concept_threshold = self.concepts_dict[concept_name]
+                result_img["concept_scores"][concept_name] = round(concept_cos - concept_threshold + adjustment,3)
```

Comment:
This means that if `adjustment > concept_threshold` -> the image will always be a bad image

Reply:
what's the theory behind this code? why you calculate the similarity between the embedding of the image and a full-ones tensor?
thanks very much